keps/sig-storage/3107-csi-nodeexpandsecret.md/3107-csi-nodeExpandSecret.md

```diff
@@ -0,0 +1,277 @@
+
+## Design Details
+
+```go
+- pkg/apis/core/types.go
+..
+type CSIPersistentVolumeSource struct {
+    .....
+    // nodeExpandSecretRef is a reference to secret object containing sensitive
+    // information to pass to the CSI driver to complete CSI node expansion
+    NodeExpandSecretRef *SecretReference
+}
+```
+The above field NodeExpandSecretRef is optional:
+
+To enable, NodeExpandSecretRef a new feature gate (NodeExpandSecret) has to be
+introduced.
+
+When the feature gate is enabled, the secretRef field will be added to the
+NodeExpandVolume request.
+
+Secrets will be fetched from StorageClass with parameters `csi.storage.k8s.io/node-expand-secret-name`
+and `csi.storage.k8s.io/node-expand-secret-namespace`. Resizing secrets will support
+same templating rules as attach and detach as documented
+- https://kubernetes-csi.github.io/docs/secrets-and-credentials.html#controller-publishunpublish-secret .
+
+CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
+field set. If not set NodeExpandVolume CSI RPC call will be made without secret.
+Existing validation of PersistentVolume object will be relaxed to allow setting of
+NodeExpandSecretRef for the first time so as CSI volume expansion can be supported
+for existing PVs.
+
+CSI Spec 1.5 has added below field to facilitate to enable COs to make use of the
+same as part of the NodeExpandSecret
+
+```
+message NodeExpandVolumeRequest {
+  ...
+  // Secrets required by plugin to complete node expand volume request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 6
+    [(csi_secret) = true, (alpha_field) = true];
+}
+```
+The same field will be used by Kubernetes to fill secretRef in the 
+NodeExpandVolume request.
+
+### Test Plan
+- Unit tests around all the added logic in kubelet.
+- E2E tests around nodeExpansionVolume to make sure the field value is passed
+  and can be used.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implemented the feature.
+- Wrote all the unit and E2E tests.
+
+#### Beta
+
+- Deployed the feature in production and went through at least minor k8s
+  version.
+
+#### GA
+
+#### Deprecation
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  - Feature gate name: NodeExpandSecret
+  - Components depending on the feature gate: kubelet, kube-apiserver
+  - Will enabling / disabling the feature require downtime of the control
+    plane? no.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? yes.
+
+- **Does enabling the feature change any default behavior?** no.
+
+- **Can the feature be disabled once it has been enabled (i.e. can we roll
+  back the enablement)?** yes, as long as the new fields in `NodeExpandRequest` is
```

Comment:
If you rollback/disable the feature when the field is set, the field will still exist but be ignored.

Reply:
sure, put that way.. thanks.